clustered/client/src/main/java/org/ehcache/clustered/client/internal/store/StrongServerStoreProxy.java

```diff
@@ -0,0 +1,197 @@
+            countDownLatch.countDown();
+          }
+        } else {
+          LOGGER.debug("CLIENT: on cache {}, ignoring invalidation on unrelated cache : {}", messageFactory.getCacheId(), response.getCacheId());
+        }
+      }
+    });
+    entity.addResponseListener(EhcacheEntityResponse.ClientInvalidateHash.class, new EhcacheClientEntity.ResponseListener<EhcacheEntityResponse.ClientInvalidateHash>() {
+      @Override
+      public void onResponse(EhcacheEntityResponse.ClientInvalidateHash response) {
+        final String cacheId = response.getCacheId();
+        final long key = response.getKey();
+        final int invalidationId = response.getInvalidationId();
+
+        LOGGER.debug("CLIENT: doing work to invalidate hash {} from cache {} (ID {})", key, cacheId, invalidationId);
+        for (InvalidationListener listener : invalidationListeners) {
+          listener.onInvalidationRequest(key);
+        }
+
+        try {
+          LOGGER.debug("CLIENT: ack'ing invalidation of hash {} from cache {} (ID {})", key, cacheId, invalidationId);
+          entity.invoke(messageFactory.clientInvalidateHashAck(key, invalidationId), true);
+        } catch (Exception e) {
+          //TODO: what should be done here?
+          LOGGER.error("error acking client invalidation of hash {} on cache {}", key, cacheId, e);
+        }
+      }
+    });
+    entity.addDisconnectionListener(new EhcacheClientEntity.DisconnectionListener() {
+      @Override
+      public void onDisconnection() {
+        for (Map.Entry<Long, CountDownLatch> entry : invalidationsInProgress.entrySet()) {
+          entry.getValue().countDown();
+        }
+        invalidationsInProgress.clear();
+      }
+    });
+  }
+
+  private <T> T performWaitingForInvalidation(long key, NullaryFunction<T> c) throws InterruptedException {
+    CountDownLatch latch = new CountDownLatch(1);
+    while (true) {
+      if (!entity.isConnected()) {
+        throw new IllegalStateException("Entity disconnected");
+      }
+      CountDownLatch countDownLatch = invalidationsInProgress.putIfAbsent(key, latch);
+      if (countDownLatch == null) {
+        break;
+      }
+      awaitOnLatch(countDownLatch);
+    }
+
+    try {
+      T result = c.apply();
+      awaitOnLatch(latch);
+      LOGGER.debug("CLIENT: key {} invalidated on all clients, unblocking append", key);
+      return result;
+    } catch (Exception ex) {
+      invalidationsInProgress.remove(key);
+      latch.countDown();
+      throw new RuntimeException(ex);
+    }
+  }
+
+  private void awaitOnLatch(CountDownLatch countDownLatch) throws InterruptedException {
+    int totalAwaitTime = 0;
+    int backoff = INVALIDATION_BACKOFF_SECONDS;
+    while (!countDownLatch.await(backoff, TimeUnit.SECONDS)) {
+      totalAwaitTime += backoff;
+      if (totalAwaitTime >= MAX_INVALIDATION_WAIT_SECONDS) {
+        throw new IllegalStateException("Server did not send an InvalidationDone message within " + totalAwaitTime + "s");
```

Comment:
I am not sure we want a max time here. In a very slow but still moving system, we may hit that limit.
I would rather have something like waiting for 1 second, then for 2 then 4 then 10 and from then on always 10. With the logging at each wake up. Makes sense?